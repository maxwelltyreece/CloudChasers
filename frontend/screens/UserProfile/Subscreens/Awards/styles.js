import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
	container: {
		justifyContent: 'flex-start',
		alignItems: 'center',
		height: '100%',
	},
	header: {
		width: '100%',
		padding: 20,
		justifyContent: 'flex-end',
		alignItems: 'flex-end',
		height: '10%',
	},
	title: {
		fontFamily: 'Montserrat_700Bold',
		fontSize: 28,
		fontWeight: 'bold',
		alignSelf: 'flex-start',
		top: '18%',
	},
	legendContainer: {
		flexDirection: 'row',
		justifyContent: 'space-around',
		alignItems: 'center',
		paddingTop: 25,
		paddingBottom: 10,
		height: '10%',
	},
	legendIcon: {
		marginRight: 5,
	},
	legendText: {
		fontFamily: 'Montserrat_500Medium',
		fontSize: 14,
	},
	awardsListContainer: {
		width: '100%',
		height: '85%',
		padding: 20,
	},
	awardItem: {
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center',
		height: '10%',
		paddingVertical: 12,
		paddingHorizontal: 20,
		backgroundColor: 'white',
		borderRadius: 12,
		marginBottom: '8%',
	},
	awardName: {
		fontFamily: 'Montserrat_700Bold',
		fontSize: 16,
		fontWeight: 'bold',
	},
	statusBubble: {
		width: 24,
		height: 24,
		borderRadius: 12,
		justifyContent: 'center',
		alignItems: 'center',
	},
	completed: {
		backgroundColor: 'green',
	},
	notCompleted: {
		backgroundColor: 'orange',
	},
	iconText: {
		fontFamily: 'Montserrat_700Bold',
		color: 'white',
		fontWeight: 'bold',
	},
});
