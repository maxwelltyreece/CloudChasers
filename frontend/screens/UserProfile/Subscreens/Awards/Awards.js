import React, { useCallback, useEffect } from 'react';
import { View, Text, ScrollView, } from 'react-native';
import PropTypes from 'prop-types';
import Icon from 'react-native-vector-icons/Feather';
import { useFocusEffect } from '@react-navigation/native';
import { useAwards } from '../../../../contexts/AwardsContext';
import { styles } from './styles';

/**
 * Awards component
 * @component
 */
const Awards = () => {
	const { userAwards, awards, fetchUserAwards, fetchAwards, fetchAwardsToBeIssued } = useAwards();

	/**
     * Fetches user awards and awards on component mount
     */
	fetchAwardsToBeIssued();

	useEffect(() => {
		fetchUserAwards();
		fetchAwards();
	}, []);

	/**
     * Fetches user awards and awards on component focus
     */
	fetchAwardsToBeIssued();

	useFocusEffect(
		useCallback(() => {
			fetchUserAwards();
			fetchAwards();
		}, [])
	);

	/**
     * Checks if an award is completed
     * @param {string} awardId - The ID of the award
     * @returns {boolean} - True if the award is completed, false otherwise
     */
	const isAwardCompleted = (awardId) => {
		return userAwards.some(((userAward) => userAward.personalAwardID === awardId));
	};

	return (
		<ScrollView contentContainerStyle={styles.container}>
			<View style={styles.legendContainer}>
				<Icon name="check-circle" size={24} color="green" style={styles.legendIcon} />
				<Text style={styles.legendText}>Completed</Text>
				<Icon name="circle" size={24} color="#FF815E" style={[styles.legendIcon, { marginLeft: 32 }]} />
				<Text style={styles.legendText}>Not Completed</Text>
			</View>
			<View style={styles.awardsListContainer}>
				{awards.map((award) => (
					<View key={award._id} style={styles.awardItem}>
						<Text style={styles.awardName} numberOfLines={1}>{award.name}</Text>
						<View style={styles.doneCheckSection} testID='icon-section'>
							{isAwardCompleted(award._id) ? (
								<Icon name="check-circle" size={24} color="green" testID='completed-icon'/>
							) : (
								<Icon name="circle" size={24} color="#FF815E" testID='not-completed-icon'/>
							)}
						</View>
					</View>
				))}
			</View>
		</ScrollView>
	);
};

export default Awards;