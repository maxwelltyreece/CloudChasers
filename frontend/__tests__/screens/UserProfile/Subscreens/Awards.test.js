/* eslint-disable no-undef */
import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import Awards from '../../../../screens/UserProfile/Subscreens/Awards/Awards';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
// import { AwardsContextProvider } from '../../../../contexts/AwardsContext';

jest.mock('@react-navigation/native', () => {
	return {
		...jest.requireActual('@react-navigation/native'),
		useFocusEffect: jest.fn(),
	};
});

// Mocking the AwardsContext
jest.mock('../../../../contexts/AwardsContext', () => ({
    useAwards: () => ({
        userAwards: mockUserAwards,
        awards: mockAwards,
        fetchUserAwards: jest.fn(),
        fetchAwards: jest.fn(),
        fetchAwardsToBeIssued: jest.fn(),
    }),
}));


const mockUserAwards = [
	{ "__v": 0, "_id": "2", "description": "Log in 5 days in a row", "name": "5 Day Streak" },
	{ "__v": 0, "_id": "4", "description": "Join a community to win this award", "name": "Join Community" },
];
const mockAwards = [
	{ "__v": 0, "_id": "1", "description": "Log in 10 days in a row", "name": "10 Day Streak" },
	{ "__v": 0, "_id": "2", "description": "Log in 5 days in a row", "name": "5 Day Streak" },
	{ "__v": 0, "_id": "3", "description": "Log in 25 days in a row", "name": "25 Day Streak" },
	{ "__v": 0, "_id": "4", "description": "Join a community to win this award", "name": "Join Community" },
	{ "__v": 0, "_id": "5", "description": "Make a post to win this award", "name": "Make a Post" },
	{ "__v": 0, "_id": "6", "description": "Make 5 posts to win this award", "name": "Make 5 Posts" },
	{ "__v": 0, "_id": "7", "description": "Make 10 posts to win this award", "name": "Make 10 Posts" }
];


describe('Awards Screen', () => {
	const Stack = createStackNavigator();

	const renderComponent = () =>
		render(
			<NavigationContainer>
				<Stack.Navigator>
					<Stack.Screen name="Awards" component={Awards} />
				</Stack.Navigator>
			</NavigationContainer>
		);

	it('renders correctly', () => {
		const { getByText } = renderComponent();

		expect(getByText('Completed')).toBeTruthy();
		expect(getByText('Not Completed')).toBeTruthy();
	});

	it('matches the awards snapshot', () => {
		const tree = renderComponent().toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('dislays the awards', () => {
		const { getByText } = renderComponent();

		expect(getByText('5 Day Streak')).toBeTruthy();
		expect(getByText('Join Community')).toBeTruthy();
		expect(getByText('10 Day Streak')).toBeTruthy();
		expect(getByText('25 Day Streak')).toBeTruthy();
		expect(getByText('Make a Post')).toBeTruthy();
		expect(getByText('Make 5 Posts')).toBeTruthy();
		expect(getByText('Make 10 Posts')).toBeTruthy();
	});

	// it('displays the awards that have been completed', () => {
	//     const { getByText, queryAllByTestId } = renderComponent();

	//     expect(getByText('5 Day Streak')).toBeTruthy();
	//     expect(getByText('Join Community')).toBeTruthy();
	//     expect(queryAllByTestId('icon-section').length).toBe(7);

	// });
    
});

