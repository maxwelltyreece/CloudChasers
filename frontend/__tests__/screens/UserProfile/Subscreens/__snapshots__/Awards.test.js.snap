// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Awards Screen matches the awards snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  Awards
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          gestureResponseDistance={
            {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          pointerEvents="box-none"
          sheetAllowedDetents="large"
          sheetCornerRadius={-1}
          sheetExpandsWhenScrolledToEdge={true}
          sheetGrabberVisible={false}
          sheetLargestUndimmedDetent="all"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            collapsable={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              [
                {
                  "display": "flex",
                  "overflow": undefined,
                },
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              {
                "close": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                handlerTag={2}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  {
                    "flex": 1,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    [
                      {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      [
                        {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <RCTScrollView
                        contentContainerStyle={
                          {
                            "alignItems": "center",
                            "height": "100%",
                            "justifyContent": "flex-start",
                          }
                        }
                      >
                        <View>
                          <View
                            style={
                              {
                                "alignItems": "center",
                                "flexDirection": "row",
                                "height": "10%",
                                "justifyContent": "space-around",
                                "paddingBottom": 10,
                                "paddingTop": 25,
                              }
                            }
                          >
                            <Text
                              allowFontScaling={false}
                              selectable={false}
                              style={
                                [
                                  {
                                    "color": "green",
                                    "fontSize": 24,
                                  },
                                  {
                                    "marginRight": 5,
                                  },
                                  {
                                    "fontFamily": "feather",
                                    "fontStyle": "normal",
                                    "fontWeight": "normal",
                                  },
                                  {},
                                ]
                              }
                            >
                              
                            </Text>
                            <Text
                              style={
                                {
                                  "fontSize": 14,
                                }
                              }
                            >
                              Completed
                            </Text>
                            <Text
                              allowFontScaling={false}
                              selectable={false}
                              style={
                                [
                                  {
                                    "color": "#FF815E",
                                    "fontSize": 24,
                                  },
                                  [
                                    {
                                      "marginRight": 5,
                                    },
                                    {
                                      "marginLeft": 32,
                                    },
                                  ],
                                  {
                                    "fontFamily": "feather",
                                    "fontStyle": "normal",
                                    "fontWeight": "normal",
                                  },
                                  {},
                                ]
                              }
                            >
                              
                            </Text>
                            <Text
                              style={
                                {
                                  "fontSize": 14,
                                }
                              }
                            >
                              Not Completed
                            </Text>
                          </View>
                          <View
                            style={
                              {
                                "height": "85%",
                                "padding": 20,
                                "width": "100%",
                              }
                            }
                          >
                            <View
                              style={
                                {
                                  "alignItems": "center",
                                  "backgroundColor": "white",
                                  "borderRadius": 12,
                                  "flexDirection": "row",
                                  "height": "10%",
                                  "justifyContent": "space-between",
                                  "marginBottom": "8%",
                                  "paddingHorizontal": 20,
                                  "paddingVertical": 12,
                                }
                              }
                            >
                              <Text
                                numberOfLines={1}
                                style={
                                  {
                                    "fontSize": 16,
                                    "fontWeight": "bold",
                                  }
                                }
                              >
                                10 Day Streak
                              </Text>
                              <View
                                testID="icon-section"
                              >
                                <Text
                                  allowFontScaling={false}
                                  selectable={false}
                                  style={
                                    [
                                      {
                                        "color": "#FF815E",
                                        "fontSize": 24,
                                      },
                                      undefined,
                                      {
                                        "fontFamily": "feather",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      {},
                                    ]
                                  }
                                  testID="not-completed-icon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                            <View
                              style={
                                {
                                  "alignItems": "center",
                                  "backgroundColor": "white",
                                  "borderRadius": 12,
                                  "flexDirection": "row",
                                  "height": "10%",
                                  "justifyContent": "space-between",
                                  "marginBottom": "8%",
                                  "paddingHorizontal": 20,
                                  "paddingVertical": 12,
                                }
                              }
                            >
                              <Text
                                numberOfLines={1}
                                style={
                                  {
                                    "fontSize": 16,
                                    "fontWeight": "bold",
                                  }
                                }
                              >
                                5 Day Streak
                              </Text>
                              <View
                                testID="icon-section"
                              >
                                <Text
                                  allowFontScaling={false}
                                  selectable={false}
                                  style={
                                    [
                                      {
                                        "color": "#FF815E",
                                        "fontSize": 24,
                                      },
                                      undefined,
                                      {
                                        "fontFamily": "feather",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      {},
                                    ]
                                  }
                                  testID="not-completed-icon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                            <View
                              style={
                                {
                                  "alignItems": "center",
                                  "backgroundColor": "white",
                                  "borderRadius": 12,
                                  "flexDirection": "row",
                                  "height": "10%",
                                  "justifyContent": "space-between",
                                  "marginBottom": "8%",
                                  "paddingHorizontal": 20,
                                  "paddingVertical": 12,
                                }
                              }
                            >
                              <Text
                                numberOfLines={1}
                                style={
                                  {
                                    "fontSize": 16,
                                    "fontWeight": "bold",
                                  }
                                }
                              >
                                25 Day Streak
                              </Text>
                              <View
                                testID="icon-section"
                              >
                                <Text
                                  allowFontScaling={false}
                                  selectable={false}
                                  style={
                                    [
                                      {
                                        "color": "#FF815E",
                                        "fontSize": 24,
                                      },
                                      undefined,
                                      {
                                        "fontFamily": "feather",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      {},
                                    ]
                                  }
                                  testID="not-completed-icon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                            <View
                              style={
                                {
                                  "alignItems": "center",
                                  "backgroundColor": "white",
                                  "borderRadius": 12,
                                  "flexDirection": "row",
                                  "height": "10%",
                                  "justifyContent": "space-between",
                                  "marginBottom": "8%",
                                  "paddingHorizontal": 20,
                                  "paddingVertical": 12,
                                }
                              }
                            >
                              <Text
                                numberOfLines={1}
                                style={
                                  {
                                    "fontSize": 16,
                                    "fontWeight": "bold",
                                  }
                                }
                              >
                                Join Community
                              </Text>
                              <View
                                testID="icon-section"
                              >
                                <Text
                                  allowFontScaling={false}
                                  selectable={false}
                                  style={
                                    [
                                      {
                                        "color": "#FF815E",
                                        "fontSize": 24,
                                      },
                                      undefined,
                                      {
                                        "fontFamily": "feather",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      {},
                                    ]
                                  }
                                  testID="not-completed-icon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                            <View
                              style={
                                {
                                  "alignItems": "center",
                                  "backgroundColor": "white",
                                  "borderRadius": 12,
                                  "flexDirection": "row",
                                  "height": "10%",
                                  "justifyContent": "space-between",
                                  "marginBottom": "8%",
                                  "paddingHorizontal": 20,
                                  "paddingVertical": 12,
                                }
                              }
                            >
                              <Text
                                numberOfLines={1}
                                style={
                                  {
                                    "fontSize": 16,
                                    "fontWeight": "bold",
                                  }
                                }
                              >
                                Make a Post
                              </Text>
                              <View
                                testID="icon-section"
                              >
                                <Text
                                  allowFontScaling={false}
                                  selectable={false}
                                  style={
                                    [
                                      {
                                        "color": "#FF815E",
                                        "fontSize": 24,
                                      },
                                      undefined,
                                      {
                                        "fontFamily": "feather",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      {},
                                    ]
                                  }
                                  testID="not-completed-icon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                            <View
                              style={
                                {
                                  "alignItems": "center",
                                  "backgroundColor": "white",
                                  "borderRadius": 12,
                                  "flexDirection": "row",
                                  "height": "10%",
                                  "justifyContent": "space-between",
                                  "marginBottom": "8%",
                                  "paddingHorizontal": 20,
                                  "paddingVertical": 12,
                                }
                              }
                            >
                              <Text
                                numberOfLines={1}
                                style={
                                  {
                                    "fontSize": 16,
                                    "fontWeight": "bold",
                                  }
                                }
                              >
                                Make 5 Posts
                              </Text>
                              <View
                                testID="icon-section"
                              >
                                <Text
                                  allowFontScaling={false}
                                  selectable={false}
                                  style={
                                    [
                                      {
                                        "color": "#FF815E",
                                        "fontSize": 24,
                                      },
                                      undefined,
                                      {
                                        "fontFamily": "feather",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      {},
                                    ]
                                  }
                                  testID="not-completed-icon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                            <View
                              style={
                                {
                                  "alignItems": "center",
                                  "backgroundColor": "white",
                                  "borderRadius": 12,
                                  "flexDirection": "row",
                                  "height": "10%",
                                  "justifyContent": "space-between",
                                  "marginBottom": "8%",
                                  "paddingHorizontal": 20,
                                  "paddingVertical": 12,
                                }
                              }
                            >
                              <Text
                                numberOfLines={1}
                                style={
                                  {
                                    "fontSize": 16,
                                    "fontWeight": "bold",
                                  }
                                }
                              >
                                Make 10 Posts
                              </Text>
                              <View
                                testID="icon-section"
                              >
                                <Text
                                  allowFontScaling={false}
                                  selectable={false}
                                  style={
                                    [
                                      {
                                        "color": "#FF815E",
                                        "fontSize": 24,
                                      },
                                      undefined,
                                      {
                                        "fontFamily": "feather",
                                        "fontStyle": "normal",
                                        "fontWeight": "normal",
                                      },
                                      {},
                                    ]
                                  }
                                  testID="not-completed-icon"
                                >
                                  
                                </Text>
                              </View>
                            </View>
                          </View>
                        </View>
                      </RCTScrollView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
